components/wifi/esp8266-driver/ESP8266Interface.cpp

```diff
@@ -351,41 +377,102 @@ int ESP8266Interface::set_channel(uint8_t channel)
 }
 
 
-int ESP8266Interface::disconnect()
+void ESP8266Interface::_disconnect_async()
 {
     _cmutex.lock();
+    _disconnect_retval = _esp.disconnect() ? NSAPI_ERROR_OK : NSAPI_ERROR_DEVICE_ERROR;
+    int timeleft_ms = ESP8266_INTERFACE_CONNECT_TIMEOUT_MS - _conn_timer.read_ms();
+
+    if (_disconnect_retval == NSAPI_ERROR_OK || ((_if_blocking == true) && (timeleft_ms <= 0))) {
+
+        if (timeleft_ms <= 0 && _connect_retval != NSAPI_ERROR_OK) {
+            _disconnect_retval = NSAPI_ERROR_CONNECTION_TIMEOUT;
+        } else {
+            if (_conn_stat != NSAPI_STATUS_DISCONNECTED) {
+                _conn_stat = NSAPI_STATUS_DISCONNECTED;
+            }
+            // In case the status update arrives later inform upper layers manually
+            _disconnect_event_id = 0;
+            _conn_timer.stop();
+            _connect_retval = NSAPI_ERROR_NO_CONNECTION;
+        }
+        // Power down the modem
+        _rst_pin.rst_assert();
```

Comment:
Having two separate pins to power down the modem is already confusing enough, so these two lines should be pushed on their own function.

Reply:
Something like a private `ESP8266Interface::_power_off()`?
`off` seems to be the stronger version of `down`, so let's use it for a function that would do both actions.
Or should we be explicit with `_power_down_and_off()`?

Reply:
I really don't see the semantic difference between those two - so _power_off() looks good to me. Function's comment section could include the explanation what is the difference between these two methods.